### ISSUE TITLE
OpenAPI on the site doesn't fully generate client and types w/ https://github.com/apple/swift-openapi-generator

### ISSUE THREAD

**Issue by heckj** [SwiftPackageIndex/SwiftPackageIndex-Server#2746]
**Please describe the bug**

Admittedly, not a priority - but since I went down the rabbit hole, I thought it worth reporting. I talked myself into trying out the Swift OpenAPI generator setup. While doing so, I worked off both the production OpenAPI.json definition, and the one in staging. Both are exhibiting errors when attempting to generate types and code from them. I figure the "staging" version is probably more relevant and up-to-date, so I'll focus there.

The error from the generator:

```
/Users/heckj/src/SPIAPI/Sources/SPIAPI/staging.openapi.json: error: Found neither a $ref nor a PathItem in Document.paths['/api/builds/{id}/doc-report'].

PathItem could not be decoded because:
Expected to find `responses` key for the **POST** endpoint under `/api/builds/{id}/doc-report` but it is missing..%
```

**Explain the steps needed to reproduce the bug**

1. Clone the generator to use it directly: `git clone https://github.com/apple/swift-openapi-generator`
2. Make a sample/example area: `mkdir ~/example`
3. Download the staging openapi file from the service into ~/example/staging.openapi.json: `curl https://staging.swiftpackageindex.com/openapi/openapi.json -o ~/example/openapi.json`
4. from within the swift-openapi-generator directory, run the generator: `swift run swift-openapi-generator generate --mode types --mode client --output-directory ~/example ~/example/openapi.json`
5. see error:

```
/Users/heckj/src/SPIAPI/Sources/SPIAPI/staging.openapi.json: error: Found neither a $ref nor a PathItem in Document.paths['/api/builds/{id}/doc-report'].

PathItem could not be decoded because:
Expected to find `responses` key for the **POST** endpoint under `/api/builds/{id}/doc-report` but it is missing..
```

**What was the expected behaviour?**

Thought I'd get a splattering of files that represented the API.

**Additional context**

This was using swift-openapi-generator on the `main` branch, commit: 23073d478d470c6e21910e84499dd1ceaf8da170 from Mon, Nov 27th, and the OpenAPI.json from staging collected November 28th. The DEV ALERTS in the Discord channel would indicate this is version 2.94.1, hash: 5bfcb82586ce0dafd811f22b3a4a57e2d78dadbc ([linked gitlab job](https://gitlab.com/finestructure/swiftpackageindex-dev/-/jobs/5621969773) that I can't access, but might have something useful) 

The error using the openapi file from production currently has a slightly different error, complaining about a missing PathItem type (I think):
```
error: Found neither a $ref nor a PathItem in Document.paths['/api/versions/{id}/build-report'].

PathItem could not be decoded because:
Expected to find `responses` key for the **POST** endpoint under `/api/versions/{id}/build-report` but it is missing..
```

CAVEAT: I haven't used OpenAPI in anger in quite a while, so I may be doing something incredibly dumb here. I started doing this as a package to import and use, just to see where I could get to, ran into errors getting even the basic package to be created, did this debugging, ... and now I'm here. The package is on GitHub ([SPIAPI](https://github.com/heckj/SPIAPI) - cause I'm awesome at naming experiments) If you're curious about what I tackled - it's a slight variation (a library package rather than a CLI) from the ["create a client" tutorial](https://swiftpackageindex.com/apple/swift-openapi-generator/0.3.5/tutorials/swift-openapi-generator/clientswiftpm).

**Comment by finestructure** [comment]
Yes, I noticed that too when I first tried the generator. I spoke to Honza about it and he recommended not to generate the spec but write the write by hand and then generate both server and client interfaces from it.

I think the problem is fixable by adding the `response:` field to the two API methods that lack it rather than rewrite the whole thing 😅

It's a quick fix and I'll deploy it to dev for testing in a minute.

**Comment by finestructure** [comment]
NB: This is not entirely fixed but it _does_ now create a `Client.swift` file. There are a few remaining problems with it though. First, it errors out when trying to generate the doc report endpoint:

```
Writing data to file Client.swift...
Error: Failed to look up a JSON Reference. 'ArrayProductDependency' was not found in schemas.
```

I'm guessing that affects build and doc reports which are at the end of the list. I believe the problem is that `ProductDependency` is self-referential and the schema doesn't reflect that:

```swift
public struct ProductDependency: Codable, Equatable {
    public var identity: String
    public var name: String
    public var url: String
    public var dependencies: [ProductDependency]
```

I don't know how to fix that.

The other problem is aesthetic - the methods names are rather ugly, `func postApiVersionsByIdBuild_hyphen_report` for instance. I'm sure that can be overridden within VaporToOpenAPI.

**Comment by heckj** [comment]
I grabbed the openAPI and run it through the validator I found at https://apitools.dev/swagger-parser/online/ - (It wouldn't grab it for me, I had to collect and paste the content as text), and I got the same error:

<img width="1001" alt="Screenshot 2023-11-29 at 10 15 06 AM" src="https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/43388/d71db8ed-26b0-4cf5-b317-5283be04705e">

My guess is that's something that's missing (or maybe broken?) from how the current OpenAPI mechanism that dynamically builds the json from your code, but hard to say - I've not looked at any of the internals there at all. Heh - and I'm not at all surprised about Honza's advice of "OpenAPI first, code from that" in really using it. I ultimately had the same take-away for some of my projects when I first started using this when it was swagger (before OpenAPI was a *thing*), but alone version 3.0.0, which I'm still trying to work through all the changes and updates. And even then, yeah - the function names were deadly ugly, even with NodeJS apps (where I was using it at the time). Swift's strong typing can't make that part any easier.

Thanks for taking a stab to further it! I'll probably play with it some more later and see what I can sort out.

**Comment by heckj** [comment]
Oooh - maybe my timing is just crap. I just saw that [OpenAPI 1.0.0 - alpha.1 was released](https://forums.swift.org/t/swift-openapi-generator-1-0-0-alpha-1-release-candidate-released-multipart-base64-filtering-recursive-types-and-more/68703), among it the feature capability for [recursive types](https://github.com/apple/swift-openapi-generator/pull/330) - exactly what you're describing.

**Comment by finestructure** [comment]
Yeah, I saw that too but I wonder if this isn't something that VaporToOpenAPI (or rather the underlying SwiftOpenAPI) needs to fix. I ran my test with yesterday's main of OpenAPI and got that failure, which I'd guess is close to what 1.0.0-alpha1 is?

**Comment by heckj** [comment]
I dug it down further, and was going to ask a question/open an issue with VaporToOpenAPI. Once I sat down with the spec that was generated and compared it to the types, the error is happening in one specific location, and it looks like it might be just a limitation of VaporToOpenAPI - how it transforms Codable types into spec. There's a recursive type structure above (`ProductDependency`] that's exposed in the DTO object for posting back build/package information from the builder, and where in some cases an array of a type is exposed properly, at the point where it goes recursive (the type contains an array of other instances like itself), that got exposed as the type `ArrayProductDependency`.

One option for a super-quick fix was just to remove that endpoint from the OpenAPI spec, but that's sort of admitting defeat 😞. VaporToOpenAPI also has some escape hatches built in - in particular customizing what it returns by conforming types to OpenAPIDescriptable to override what the default mechanisms creates, or OpenAPIType to provide a specific type with schema. Figured it was worth an ask.
